File: News_Service_Jooq/src/main/java/com/dunice/projectNews/errors/ControllerAdvice.java
//package com.dunice.projectNews.errors;
//
//import com.dunice.projectNews.dto.common.BaseSuccessResponse;
//import com.dunice.projectNews.exception.CustomException;
//import com.dunice.projectNews.util.AppConstants;
//import com.dunice.projectNews.util.ServerErrorCodes;
//import jakarta.servlet.http.HttpServletRequest;
//import jakarta.validation.ConstraintViolation;
//import jakarta.validation.ConstraintViolationException;
//import org.springframework.context.support.DefaultMessageSourceResolvable;
//import org.springframework.http.HttpStatus;
//import org.springframework.http.ResponseEntity;
//import org.springframework.web.bind.MethodArgumentNotValidException;
//import org.springframework.web.bind.annotation.ExceptionHandler;
//import org.springframework.web.bind.annotation.RestControllerAdvice;
//import org.springframework.web.multipart.MultipartException;
//import java.util.List;
//import java.util.stream.Collectors;
//
//@RestControllerAdvice
//public class ControllerAdvice {
//    @ExceptionHandler(MethodArgumentNotValidException.class)
//    public ResponseEntity<BaseSuccessResponse> handleException(MethodArgumentNotValidException e, HttpServletRequest request) {
//        List<Integer> result = e.getBindingResult().getAllErrors()
//                .stream()
//                .map(m -> ServerErrorCodes.map.get(m.getDefaultMessage())).toList();
//        request.setAttribute(AppConstants.ERRORS, e
//                .getBindingResult()
//                .getAllErrors()
//                .stream()
//                .map(DefaultMessageSourceResolvable::getDefaultMessage)
//                .collect(Collectors.joining(" ")));
//        BaseSuccessResponse baseSuccessResponse = new BaseSuccessResponse(result.get(0), true);
//        baseSuccessResponse.setCodes(result);
//        return ResponseEntity.badRequest().body(baseSuccessResponse);
//    }
//
//    @ExceptionHandler(CustomException.class)
//    public ResponseEntity<BaseSuccessResponse> handleException(CustomException e, HttpServletRequest request) {
//        BaseSuccessResponse baseSuccessResponse =
//                new BaseSuccessResponse(ServerErrorCodes.map.get(e.getMessage()), true);
//        request.setAttribute(AppConstants.ERRORS, e.getMessage());
//        return ResponseEntity.badRequest().body(baseSuccessResponse);
//    }
//
//    @ExceptionHandler(ConstraintViolationException.class)
//    public ResponseEntity<BaseSuccessResponse> handleException(ConstraintViolationException ex, HttpServletRequest request) {
//
//        List<Integer> result = ex.getConstraintViolations()
//                .stream()
//                .map(e -> ServerErrorCodes.map.get(e.getMessage()))
//                .toList();
//        request.setAttribute(AppConstants.ERRORS, ex
//                .getConstraintViolations().stream()
//                .map(ConstraintViolation::getMessage)
//                .collect(Collectors.joining(" ")));
//        return new ResponseEntity<>(new BaseSuccessResponse(result.get(0)), HttpStatus.BAD_REQUEST);
//    }
//
//    @ExceptionHandler(MultipartException.class)
//    public ResponseEntity<BaseSuccessResponse> handleException(MultipartException exception, HttpServletRequest request) {
//        BaseSuccessResponse baseSuccessResponse = new BaseSuccessResponse();
//        baseSuccessResponse.setCodes(List.of(ServerErrorCodes.map.get(ServerErrorCodes.UNKNOWN.getMessage())));
//        request.setAttribute(AppConstants.ERRORS, ServerErrorCodes.UNKNOWN.getMessage());
//        return ResponseEntity.badRequest().body(baseSuccessResponse);
//    }
//
//    @ExceptionHandler(Exception.class)
//    public ResponseEntity<BaseSuccessResponse> handleException(Exception exception, HttpServletRequest request) {
//        BaseSuccessResponse baseSuccessResponse = new BaseSuccessResponse();
//        baseSuccessResponse.setCodes(List.of(ServerErrorCodes.map.get(ServerErrorCodes.UNKNOWN.getMessage())));
//        request.setAttribute(AppConstants.ERRORS, ServerErrorCodes.UNKNOWN.getMessage());
//        return ResponseEntity.badRequest().body(baseSuccessResponse);
//    }
//}

File: News_Service_Kotlin/src/main/kotlin/com/dunice/advancedServiceKotlin/errors/ControllerAdvice.kt
package com.dunice.advancedServiceKotlin.errors

import com.dunice.advancedServiceKotlin.dto.common.BaseSuccessResponse
import com.dunice.advancedServiceKotlin.exception.CustomException
import com.dunice.advancedServiceKotlin.util.AppConstants
import com.dunice.advancedServiceKotlin.util.ServerErrorCodes
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.ConstraintViolation
import jakarta.validation.ConstraintViolationException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.ObjectError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.multipart.MultipartException
import java.util.List
import java.util.stream.Collectors

@RestControllerAdvice
class ControllerAdvice {

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleException(
        e: MethodArgumentNotValidException,
        request: HttpServletRequest
    ): ResponseEntity<BaseSuccessResponse> {

        val result = e.bindingResult.allErrors
            .map { it: ObjectError -> ServerErrorCodes.map[it.defaultMessage] }

        request.setAttribute(
            AppConstants.ERRORS,
            e.bindingResult.allErrors
                .joinToString(" ") { it.defaultMessage ?: "" }
        )

        val baseSuccessResponse = BaseSuccessResponse(result.firstOrNull() ?: 0, true)
        baseSuccessResponse.codes = result

        return ResponseEntity.badRequest().body(baseSuccessResponse)
    }


    @ExceptionHandler(CustomException::class)
    fun handleException(e: CustomException, request: HttpServletRequest): ResponseEntity<BaseSuccessResponse> {
        val baseSuccessResponse =
            BaseSuccessResponse(ServerErrorCodes.map[e.message]!!, true)
        request.setAttribute(AppConstants.ERRORS, e.message)
        return ResponseEntity.badRequest().body(baseSuccessResponse)
    }

    @ExceptionHandler(ConstraintViolationException::class)
    fun handleException(
        ex: ConstraintViolationException,
        request: HttpServletRequest
    ): ResponseEntity<BaseSuccessResponse> {
        val result = ex.constraintViolations
            .stream()
            .map { e: ConstraintViolation<*> -> ServerErrorCodes.map[e.message] }
            .toList()
        request.setAttribute(AppConstants.ERRORS, ex
            .constraintViolations.stream()
            .map { obj: ConstraintViolation<*> -> obj.message }
            .collect(Collectors.joining(" ")))
        return ResponseEntity(BaseSuccessResponse(result[0]!!), HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(MultipartException::class)
    fun handleException(
        exception: MultipartException?,
        request: HttpServletRequest
    ): ResponseEntity<BaseSuccessResponse> {
        val baseSuccessResponse = BaseSuccessResponse()
        baseSuccessResponse.codes = List.of(ServerErrorCodes.map[ServerErrorCodes.UNKNOWN.message])
        request.setAttribute(AppConstants.ERRORS, ServerErrorCodes.UNKNOWN.message)
        return ResponseEntity.badRequest().body(baseSuccessResponse)
    }

    @ExceptionHandler(Exception::class)
    fun handleException(exception: Exception?, request: HttpServletRequest): ResponseEntity<BaseSuccessResponse> {
        val baseSuccessResponse = BaseSuccessResponse()
        baseSuccessResponse.codes = List.of(ServerErrorCodes.map[ServerErrorCodes.UNKNOWN.message])
        request.setAttribute(AppConstants.ERRORS, ServerErrorCodes.UNKNOWN.message)
        return ResponseEntity.badRequest().body(baseSuccessResponse)
    }
}
